test/models/user_test.rb

```diff
@@ -990,4 +990,37 @@ def editing_self_helper
       end
     end
   end
+
+  describe '#visible_environments' do
+    test 'should check for environments permission with orgs/locs disabled' do
+      Taxonomy.expects(:locations_enabled).returns(false)
+      Taxonomy.expects(:organizations_enabled).returns(false)
+      setup_user 'view', 'environments', 'name != production'
+      all_but_production = Environment.unscoped.select { |e| e.name != 'production' }
+      assert_equal all_but_production.map(&:name), User.current.visible_environments
+    end
+
+    test 'should show the list of environments visible when only orgs are enabled' do
+      Taxonomy.expects(:locations_enabled).returns(false).at_least_once
+      setup_user 'view', 'environments', 'name = test_env_org1'
+      FactoryGirl.create(:environment, :name => 'test_env_org1',
+                         :organizations => [User.current.organizations.first],
+                         :locations => User.current.locations.all)
+      assert_equal ['test_env_org1'], User.current.visible_environments
+    end
+
+    test 'should show the list of environments visible when only locs are enabled' do
+      # No need to expect 'organizations_enabled' here - as long as locations_enabled
+      # is true, #visible_environments will check by default on taxable_taxonomies.
```

Comment:
https://github.com/theforeman/foreman/pull/4510/files#diff-4676c008b11a5480d73d4a6de01e45b9R404 - it's just that the if condition to add in 'taxonomy checks' is `locations_enabled || organizations_enabled`. When 'locations_enabled' is true, like in this case, 'organizations_enabled' will never get called.